mods/fire/init.lua

```diff
@@ -103,6 +106,28 @@ minetest.register_craft({
 	}
 })
 
+
+-- Override coalblock to enable lighting a permanent flame above
+
+minetest.override_item("default:coalblock", {
+	on_punch = function(pos, node, puncher)
+		local above = {x = pos.x, y = pos.y + 1, z = pos.z}
+		if puncher:get_wielded_item():get_name() == "default:torch" and
+				minetest.get_node(above).name == "air" then
+			minetest.after(0.2, function()
+				minetest.set_node(above, {name = "fire:permanent_flame"})
```

Comment:
If you don't search for air directly before setting the permanent_flame, the player could have placed something on the coal block, e.g. a torch, which gets removed.
You could remove searching for air before executing minetest.after because a player doesn't punch a coalblock extremely often to fill the memory with the minetest.after table.